File: agent/src/repos/mru_index.ts
// Path: agent/src/repos/mru_index.ts
// Description: MRU (Most Recently Used) index for recent file changes with unique-by-path semantics

import type { FileEntry } from "../../../app/src/shared/protocol.js";

/**
 * MRU index that maintains unique entries by path with newest-first ordering.
 * Uses Map's insertion order to track recency - delete+insert moves to end.
 */
export interface MruIndex {
  /** Add or update entry; moves to front (most recent) */
  upsert(entry: FileEntry): void;
  /** Remove entry by path; returns true if found */
  remove(path: string): boolean;
  /** Get all entries newest-first, trimmed to capacity */
  entries(): FileEntry[];
  /** Seed the index from persisted entries (oldest-first order expected) */
  loadFrom(entries: FileEntry[]): void;
  /** Current entry count */
  readonly size: number;
  /** Clear all entries */
  clear(): void;
}

/**
 * Create an MRU index with bounded capacity.
 * Entries are unique by path; updating an entry moves it to the front.
 */
export function createMruIndex(capacity: number): MruIndex {
  if (capacity <= 0) {
    throw new Error("MruIndex capacity must be positive");
  }

  // Map maintains insertion order; we treat end as "newest"
  const items = new Map<string, FileEntry>();

  function upsert(entry: FileEntry): void {
    // Delete first to update insertion order (move to end = newest)
    items.delete(entry.path);
    items.set(entry.path, entry);

    // Trim oldest entries if over capacity
    while (items.size > capacity) {
      const firstKey = items.keys().next().value as string;
      items.delete(firstKey);
    }
  }

  function remove(path: string): boolean {
    return items.delete(path);
  }

  function entries(): FileEntry[] {
    // Convert to array and reverse so newest is first
    return Array.from(items.values()).reverse();
  }

  function loadFrom(persistedEntries: FileEntry[]): void {
    items.clear();
    // Persisted entries should be newest-first, so iterate in reverse
    // to add oldest first (so newest ends up at end of Map = front of output)
    for (let i = persistedEntries.length - 1; i >= 0; i--) {
      const entry = persistedEntries[i];
      if (entry && !items.has(entry.path)) {
        items.set(entry.path, entry);
      }
    }
    // Trim to capacity
    while (items.size > capacity) {
      const firstKey = items.keys().next().value as string;
      items.delete(firstKey);
    }
  }

  function clear(): void {
    items.clear();
  }

  return {
    upsert,
    remove,
    entries,
    loadFrom,
    get size() {
      return items.size;
    },
    clear,
  };
}

File: crates/im_agent/src/repos/mru_index.rs
// Path: crates/im_agent/src/repos/mru_index.rs
// Description: MRU index for recent file changes

use crate::protocol::FileEntry;
use crate::repos::{
    activity_from_mtime, normalize_activity_history, observed_at_from_mtime, update_activity,
};

#[derive(Debug)]
pub struct MruIndex {
    capacity: usize,
    entries: Vec<FileEntry>,
}

impl MruIndex {
    pub fn new(capacity: usize) -> Result<Self, String> {
        if capacity == 0 {
            return Err("MruIndex capacity must be positive".to_string());
        }
        Ok(Self {
            capacity,
            entries: Vec::new(),
        })
    }

    pub fn upsert(&mut self, mut entry: FileEntry) -> FileEntry {
        let previous_activity = self
            .entries
            .iter()
            .find(|item| item.path == entry.path)
            .and_then(|item| item.activity.as_ref());
        let observed_at = observed_at_from_mtime(&entry.mtime);
        entry.activity = Some(update_activity(previous_activity, observed_at));

        self.entries.retain(|item| item.path != entry.path);
        self.entries.insert(0, entry);
        if self.entries.len() > self.capacity {
            self.entries.truncate(self.capacity);
        }
        self.entries[0].clone()
    }

    pub fn remove(&mut self, path: &str) -> bool {
        let before = self.entries.len();
        self.entries.retain(|item| item.path != path);
        before != self.entries.len()
    }

    pub fn entries(&self) -> Vec<FileEntry> {
        self.entries.clone()
    }

    pub fn load_from(&mut self, persisted: Vec<FileEntry>) {
        let mut entries = Vec::new();
        for mut entry in persisted {
            if entries
                .iter()
                .any(|item: &FileEntry| item.path == entry.path)
            {
                continue;
            }
            if entry.activity.is_none() {
                entry.activity = Some(activity_from_mtime(&entry.mtime));
            }
            if let Some(activity) = &mut entry.activity {
                normalize_activity_history(activity, &entry.mtime);
            }
            entries.push(entry);
            if entries.len() == self.capacity {
                break;
            }
        }
        self.entries = entries;
    }

    pub fn clear(&mut self) {
        self.entries.clear();
    }

    pub fn size(&self) -> usize {
        self.entries.len()
    }
}
